table/tables/mutation_checker.go

```diff
@@ -0,0 +1,243 @@
+	return nil
+}
+
+// checkIndexKeys checks whether the decoded data from keys of index mutations are consistent with the expected ones.
+func checkIndexKeys(sc *stmtctx.StatementContext, sessVars *variable.SessionVars, t *TableCommon,
+	dataAdded []types.Datum, dataRemoved []types.Datum, indexMutations []mutation) error {
+	indexIDMap := make(map[int64]indexHelperInfo)
+	for _, index := range t.indices {
+		if index.Meta().Primary && t.meta.IsCommonHandle {
+			continue
+		}
+		indexIDMap[index.Meta().ID] = indexHelperInfo{
+			index.Meta(),
+			BuildRowcodecColInfoForIndexColumns(index.Meta(), t.Meta()),
+		}
+	}
+
+	for _, m := range indexMutations {
+		_, indexID, _, err := tablecodec.DecodeIndexKey(m.key)
+		if err != nil {
+			//FIXME: continue or return error?
+			return errors.Trace(err)
+		}
+
+		indexHelperInfo, ok := indexIDMap[indexID]
+		if !ok {
+			return errors.New("index not found")
+		}
+
+		// when we cannot decode the key to get the original value
+		if len(m.value) == 0 && NeedRestoredData(indexHelperInfo.indexInfo.Columns, t.Meta().Columns) {
+			continue
+		}
+
+		decodedIndexValues, err := tablecodec.DecodeIndexKV(m.key, m.value, len(indexHelperInfo.indexInfo.Columns),
+			tablecodec.HandleNotNeeded, indexHelperInfo.rowColInfos)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		indexData := make([]types.Datum, 0)
+		for i, v := range decodedIndexValues {
+			fieldType := &t.Columns[indexHelperInfo.indexInfo.Columns[i].Offset].FieldType
+			datum, err := tablecodec.DecodeColumnValue(v, fieldType, sessVars.Location())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			indexData = append(indexData, datum)
+		}
+
+		if len(m.value) == 0 {
+			err = compareIndexData(sc, t.Columns, indexData, dataRemoved, indexHelperInfo.indexInfo)
+		} else {
+			err = compareIndexData(sc, t.Columns, indexData, dataAdded, indexHelperInfo.indexInfo)
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// checkRowInsertionConsistency checks whether the values of row mutations are consistent with the expected ones
+// We only check data added since a deletion of a row doesn't care about its value (and we cannot know it)
+func checkRowInsertionConsistency(sessVars *variable.SessionVars, tableColumns []*model.ColumnInfo, rowToInsert []types.Datum, rowInsertions []mutation) error {
+	if rowToInsert == nil {
+		// it's a deletion
+		return nil
+	}
+	if len(rowInsertions) == 0 {
+		return errors.New("row mutations not found for a put of a row")
+	}
+	if len(rowInsertions) > 1 {
+		// impossible to have multiple insertions of rows when modifying a single row of in a table
+		logutil.BgLogger().Error("multiple row mutations added/mutated", zap.Any("rowInsertions", rowInsertions))
+		return errors.New("multiple row mutations added/mutated")
+	}
+
+	columnMap := make(map[int64]*model.ColumnInfo)
+	for _, col := range tableColumns {
+		columnMap[col.ID] = col
+	}
```

Comment:
This looks too expensive if we do this for each row... If this is really necessary, can we try to store it somewhere like the sessionctx, to make it reusable?

Reply:
I tried to save it in the stmtctx. PTAL if it's reasonable